#32 feat : 친구 추가 및 초대 코드 생성

yunah7 opened the pull request:
## 요약
친구 추가 기능 및 관련된 예외 추가

## 변경 내용
**1. ConnectionController**
```
@RestController
@RequestMapping("/connections")
@RequiredArgsConstructor
public class ConnectionController {
	private final ConnectionService connectionService;

	@PostMapping("/get-url")
	public String getInviteUrl() {
		return connectionService.getInviteUrl();
	}

	@PostMapping("create/{connectorCode}")
	public void createConnection(@PathVariable String connectorCode) {
		connectionService.createConnection(connectorCode);
	}

}
```
- 초대 URL을 String으로 반환하는 메소드 추가 
  - 이후 프론트를 만들게 된다면 URL 자체가 화면에 출력되야 하기에 String 사용
  - 유저 정보를 가져오는 것이지만, 초대URL 자체는 connect 시에만 사용되기에 해당 컨트롤러에 추가 
- 초대 코드를 PathVariable로 받아 친구를 추가하는 메소드 생성


**2. Connection**
```
import lombok.AllArgsConstructor;
import lombok.Getter;

@Getter
@AllArgsConstructor
public class Connection {
	private int id;
	private int connectorId;
	private int connecteeId;
	boolean activate;

}
```
- CONNECTOR 테이블 내 정보들엔 중요한 개인정보가 들어있지는 않음
-  column을 찾기 위해 각각 DB에 연결하는 것이 비효율적이라 생각하여, VO 객체 생성


**3. ConnectMapper**
```
@Mapper
public interface ConnectionMapper {

	String getInviteUrlCode(int id);

	Connection getConnection(@Param("connectorId") int connectorId, @Param("connecteeId") int connecteeId);

	void modifyActivate(int id, boolean status);

	void createConnection(@Param("connectorId") int connectorId, @Param("connecteeId") int connecteeId);
}
```
- 유저의 id를 통해 초대 코드를 반환하는 메소드
- Connection 객체를 반환하는 메소드
- CONNECTOR의 activate를 변경하는 메소드
  - activate를 통해 친구 추가와 삭제를 관리, 매개변수로 boolean을 받아 옴
- connection을 생성하는 메소드 


**4. ConnectionService**
```
@Service
@RequiredArgsConstructor
public class ConnectionService {

	private final LoginService loginService;

	private final ConnectionMapper connectionMapper;

	private final UserService userService;

	public String getInviteUrl() {
		int loginUserId = loginService.getLoginSession();
		StringBuilder inviteUrl = new StringBuilder();
		inviteUrl.append("http://localhost:8080/connections/create/")
			.append(connectionMapper.getInviteUrlCode(loginUserId));
		return inviteUrl.toString();
	}

	public void createConnection(String connectorCode) {
		int loginUserId = loginService.getLoginSession();
		int connectorId = userService.getIdByInviteCode(connectorCode);
		if (connectionMapper.getInviteUrlCode(loginUserId).equals(connectorCode)) {
			throw new AuthorizationException(ErrorCode.SELF_INVITATION);
		}
		Connection connection = connectionMapper.getConnection(loginUserId, connectorId);
		if (connection != null) {
			if (connection.isActivate()) {
				throw new DuplicateException(ErrorCode.CONNECT_DUPLICATE);
			} else {
				connectionMapper.modifyActivate(connection.getId(), true);
			}
		}
		connectionMapper.createConnection(loginUserId, connectorId);
	}
}
```
- getInviteUrl
  - DB 내 초대 코드를 URL로 바꾸어 전달
- createConnection
  - 초대 링크를 준 유저(connertorId), 받은 유저(loginUserId)의 id를 얻어 옴
  - connection을 확인하기 전, 초대링크 내 초대 코드가 초대링크를 받은 유저의 초대코드와 동일한지 확인
  - 유저의 id를 통해 connection을 받아옴
  - connect 되지 않은 유저라면 바로 connection을 맺어 줌
  - connect 된 유저라면, activate 상태를 살펴 친구 삭제가 된 관계일 시, 친구 관계를 복구

   
**5. UserService**
```
	private String createInviteCode() {
		return RandomStringUtils.randomAlphabetic(12);
	}
```
- 초대 코드를 만들어주는 createInviteCode 메소드
  - 초대 코드를 통해 유저를 유추할 수 있으면 안됨 (= 아이디와 연관이 있으면 안됨)
    - 해당 프로젝트 내 친구 추가는 요청-대기-승인 체계가 아닌 요청시 바로 승인이 되는 방식
    - 차후 추가할 페이지(랜덤 유저의 일정 조회)에서 유저의 id가 공개 됨
  - 적당히 unique 해야 함
    -  영대소문자를 구별해 12글자로 이루어진 문자열 
  - 단, Random 기반으로 생성 시간을 seed로 하여 안전하다는 말할 수 없고, 겹칠 가능성이 희박하지만 없다고 할 수는 없음 


**6. ConnectionMapper**
```
    <select id="getInviteUrlCode" resultType="String">
        SELECT invite_url_code
        FROM USER_TB
        WHERE id = #{id};
    </select>

    <select id="getConnection" resultType="study.sharedcalendar.dto.Connection">
        SELECT id, connector_id, connectee_id, activate
        FROM CONNECTOR_TB
        WHERE (connector_id = #{connectorId} AND connectee_id = #{connecteeId})
           OR (connectee_id = #{connectorId} AND connector_id = #{connecteeId});
    </select>

    <update id="modifyActivate">
        UPDATE CONNECTOR_TB
        SET activate = #{status}
        WHERE id = #{id};
    </update>

    <insert id="createConnection">
        INSERT INTO CONNECTOR_TB (connector_id, connectee_id, last_visited_at, activate, created_at, updated_at)
        values (#{connectorId}, #{connecteeId}, NOW(), 1, NOW(), NOW())
    </insert>
```
- getConnection
  - connector, connectee 자리에 들어올 사람은 고정되지 않음
  - 1과 2의 connection에선 1이 connector에 1과 3의 connection에선 1이 connertee 자리에 들어갈 수 있음  


## Issue number/URL
#32

yunah7 committed:
#32 feat : 친구 추가 및 초대 코드 생성

RandomStringUtils 내 randomAlphabetic을 통한 랜덤 문자열로 초대코드 생성
PathVariable 어노테이션을 통해 초대코드로 친구 추가

yunah7 commented:
컨벤션이 제대로 적용되지 않는 것 같아 이 부분은 다시 설정하고 커밋하겠습니다!

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -0,0 +1,27 @@
+public class ConnectionController {
+	private final ConnectionService connectionService;
+
+	@PostMapping("/get-url")
+	public String getInviteUrl() {
+		return connectionService.getInviteUrl();
+	}
```
조회 api같은데 GetMapping으로 바꾸고 uri는 "/url"만 있어도 충분해보이네요 :)

yunah7 replied:
수정하겠습니다!

uri010 replied:
초대링크의 경우 친구 개념인 Connection보다는 자신의 정보를 가져오는 User에 좀 더 적합한 것 같습니다. 
고민한대로 친구 요청시에만 불러오는 거라 ConnectionController에 들어가는 것도 이해는 되지만, ConnectionMapper에서 user 테이블의 정보를 조회하는게 조금 안 맞는 것 같습니다..!

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -0,0 +1,27 @@
+		return connectionService.getInviteUrl();
+	}
+
+	@PostMapping("create/{connectorCode}")
+	public void createConnection(@PathVariable String connectorCode) {
+		connectionService.createConnection(connectorCode);
+	}
```
이것도 Post 자체에 create 의미가 포함되어 있기 때문에 create는 생략해도 될 것 같습니다. 차라리 code를 적어주는 것도 나아보이네요!

sunwootest commented on src/main/java/study/sharedcalendar/service/ConnectionService.java:
```diff
@@ -0,0 +1,45 @@
+	private final UserService userService;
+
+	public String getInviteUrl() {
+		int loginUserId = loginService.getLoginSession();
+		StringBuilder inviteUrl = new StringBuilder();
+		inviteUrl.append("http://localhost:8080/connections/create/")
+			.append(connectionMapper.getInviteUrlCode(loginUserId));
+		return inviteUrl.toString();
```
http://localhost:8080에 의존해있네요. 이를 어떻게 하면 해결할 수 있을지 고민해보면 좋겠습니다. 그리고 StringBuilder는 의도된걸까요?

sunwootest replied:
초대코드가 한번 정해지면 안변하는 것 같은데 이로인해 발생할 수 있는 문제점이 뭘까요? 또 어떤식으로 해결할 수 있을까요?

yunah7 replied:
사용자의 요청 시마다 String을 더하는 연산이 이루어져야 하기에 String을 사용하는 것은 비효율적이라고 생각하였고, 단순 조회이기에 Thread Safe가 큰 의미가 없을 것이라 생각했는데, 다시 생각해보니 JDK 1.5 이후부턴 String 클래스도 StringBuilder와 같이 처리되며, Thread safe 하지 않을 경우 append 과정에서 의도치 않은 결과가 나올 수 있다 생각해 StringBuffer로 바꾸려합니다 :)

yunah7 replied:
http://localhost:8080은 원래 메소드를 통해 URL을 받아오려 했지만, HttpServletRequest 자체에서 URL을 받아오는 방법이나, UrlPathHelper 객체를 만드는 방법 모두 request 객체를 필요로 해, String으로 넘기게 되었습니다. 😢 

혹 Constant로 두고 사용하라는 말씀이실까요, 아니면 request 객체 없이도 서비스단에서 URL을 받아올 수 있는 방법이 있다는 말씀이실까요?

yunah7 replied:
초대코드가 변하지 않으면 아무래도 보안상 문제를 가장 클 것 같습니다. DB 유출 문제가 아니더라도, 사용자의 실수로 초대 코드가 공개되는 경우, 요청-대기 단계가 없는 로직의 특성 상, 원치 않는 connection을 맺게 되기에 문제가 될 것 같습니다.

초대코드를 변하지 않도록 한 이유가 사용자의 편의성 때문이었는데, 링크로 초대링크를 전송한다는 것은 SNS를 통해 전송한다는 것이기에, 즉각적인 사용이 이루어지지 않을 가능성이 커, 임의의 시간동안 코드를 유지시키게되면 편의성이 떨어질 것이라 생각했습니다.

 초대 코드를 아예 변하지 않게 하기 보단, 사용자가 변경 요청을 할 때마다 코드가 변경되는 방식으로 변경하겠습니다 :)

sunwootest replied:
세션을 이용해서 로그인을 구현했던데 꼭 세션은 아니지만 이와 비슷한 원리를 적용하면 좋을 것 같습니다. 유효기간을 설정할 수 있거든요 :)

sunwootest replied:
다양한 방법이 있겠지만 비즈니스 로직 내부에 명확한 url을 기입하는 것은 바람직해 보이지 않아서요.

sunwootest replied:
지금 당장 이해하기 어려울 순 있지만 저 상황이 과연 동시성 이슈가 발생할 수 있는 로직인지 알아보는 것도 좋을 것 같습니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/ConnectionService.java:
```diff
@@ -0,0 +1,45 @@
+	private final ConnectionMapper connectionMapper;
+	private final UserService userService;
+
+	public String getInviteUrl() {
+		int loginUserId = loginService.getLoginSession();
+		StringBuilder inviteUrl = new StringBuilder();
+		inviteUrl.append("http://localhost:8080/connections/create/")
+			.append(connectionMapper.getInviteUrlCode(loginUserId));
+		return inviteUrl.toString();
+	}
+
+	public void createConnection(String connectorCode) {
+		int loginUserId = loginService.getLoginSession();
+		int connectorId = userService.getIdByInviteCode(connectorCode);
+		if (connectionMapper.getInviteUrlCode(loginUserId).equals(connectorCode)) {
+			throw new AuthorizationException(ErrorCode.SELF_INVITATION);
+		}
+		Connection connection = connectionMapper.getConnection(loginUserId, connectorId);
+		if (connection != null) {
+			if (connection.isActivate()) {
+				throw new DuplicateException(ErrorCode.CONNECT_DUPLICATE);
+			} else {
+				connectionMapper.modifyActivate(connection.getId(), true);
+			}
+		}
+		connectionMapper.createConnection(loginUserId, connectorId);
+	}
```
코드 사이에 개행을 넣어주면 가독성이 좋아질 것 같습니다 :)

sunwootest commented on src/main/resources/ddl.sql:
```diff
@@ -1,19 +1,19 @@
-    activate      tinyint     not null,
-    try_count     int         not null,
-    created_at    timestamp   not null default current_timestamp,
-    updated_at    timestamp   not null default current_timestamp on update current_timestamp,
-    password_date timestamp   not null default current_timestamp,
+    id              int generated by default as identity,
+    user_id         varchar(45) not null,
+    password        text        not null,
+    email           varchar(45) not null,
+    invite_url_code text        not null,
+    activate        tinyint     not null,
+    try_count       int         not null,
+    created_at      timestamp   not null default current_timestamp,
+    updated_at      timestamp   not null default current_timestamp on update current_timestamp,
+    password_date   timestamp   not null default current_timestamp,
     primary key (id)
 );
```
password가 꼭 text type이어야할까요?

yunah7 replied:
Bcrypt가 항상 길이가 60인 String을 만들어주기에 text보단 varchar이 효율적일 것 같습니다 ! 수정하겠습니다 :)

uri010 replied:
activate 같이 값이 두개로 한정되는 경우, 즉 true/false로만 나뉘는 경우는 boolean으로 처리하는 것이 좋을 것 같습니다! 
처음엔 여러 상태가 존재하는 경우를 생각해 tinyint로 설정해뒀는데 요구사항을 생각해봤을 때 boolean을 사용해도 될 것 같습니다.

sunwootest replied:
mysql 기준으로 boolean은 tinyint(1)로 표현할 수 있습니다!

yunah7 committed:
#32 fix : 로그인 - 초대 코드 변경 기능 및 상수 처리

feature/30 branch merge 후 변경 내용 연동
초대 코드 변경 메소드 추가
초대 URL 및 초대 코드 자릿수 상수로 처리
password sql 타입 varchar(60)으로 변경
controller의 mapping url 이름 수정

yunah7 commented:
## 변경 내용
**1. 컨벤션 적용**

**2. 개행 처리**

**3. merge된 feature/30 내용 반영**

**4. password sql 타입 text->varchar(60)**

**5. UserConstant**
```java
        public static String LOGIN_SESSION_ID;
	public static String USER_INVITE_URL;
	public static int USER_INVITE_CODE_SIZE;

	public UserConstant(int maxLoginTryCount, int maxPasswordValidityPeriod, String loginSessionId,
		String userInviteUrl, int userInviteCodeSize) {
		MAX_LOGIN_TRY_COUNT = maxLoginTryCount;
		MAX_PASSWORD_VALIDITY_PERIOD = maxPasswordValidityPeriod;
		LOGIN_SESSION_ID = loginSessionId;
		USER_INVITE_URL = userInviteUrl;
		USER_INVITE_CODE_SIZE = userInviteCodeSize;
	}
```
- url 및 초대 코드의 사이즈를 상수로 추가

**6. ConnectController**
```java
	@PostMapping("/url")
	public String getInviteUrl() {
		return connectionService.getInviteUrl();
	}

	@PostMapping("/change-code")
	public void changeInviteUrlCode() {
		connectionService.changeInviteCode();
	}

	@PostMapping("{connectorCode}")
	public void createConnection(@PathVariable String connectorCode) {
		connectionService.createConnection(connectorCode);
	}
```
- url명 수정
- 유저가 초대코드를 바꿀 수 있도록 메소드 추가
   - 유효시간도 고려해보았으나, 편리성면에서 사용자가 원하는 시간대에 초대 코드를 변경할 수 있도록 하는 것이 초대 링크 기능의 구현 목적에 부합하다 느껴, 해당 메소드를 추가했습니다.
- getInviteUrl 메소드 유지
   - User 서비스단에서 다루는 것이 관심사면에선 적합하지만, Url을 가져오는 기능은 로그인 된 유저만이 가능하기에, 저장된 세션에서 유저 정보를 꺼내와야 하는데, 현재 LoginService가 UserService를 참조하고 있어, UserService에 해당메소드를 넣을 경우, 순환참조 문제가 발생해 현 상태를 유지했습니다.  
   - 이후, 로그인된 유저의 정보를 가져와야 하는 기능이 추가로 구현되어야 한다면, 해당 기능끼리 묶어 분리하겠습니다.

**7. ConnectionService**
```java
        public String getInviteUrl() {
		int loginId = loginService.getLoginSession();
		return (USER_INVITE_URL + connectionMapper.getInviteUrlCode(loginId));
	}

	public void changeInviteCode() {
		int loginId = loginService.getLoginSession();
		userService.modifyInviteUrlCode(loginId);
	}
```
- URL을 상수로 처리
- 초대 코드 수정 로직 추가

## ps
로컬 커밋 실수로 해당 내용을 작업 중이던 브랜치가 4시간 가량 시도해보아도 복구가 안되어 다른 브랜치에 내용을 옮기고 push를 해 force-push 기록이 찍히게 되었습니다..!
force-push로도 내용 변경이 되지 않은 것 같아, feature/32 branch를 리셋하고 다시 내용을 옮겨 정상적으로 push 하였고, 결론적으로도 force-push가 아닌 정상 push와 관련된 커밋만 적용된 것 같아 현 상태로 comment 올립니다.

uri010 commented on src/main/java/study/sharedcalendar/interceptor/AutoLoginInterceptor.java:
```diff
@@ -18,15 +18,15 @@
 @RequiredArgsConstructor
 public class AutoLoginInterceptor implements HandlerInterceptor {
-
 	private final UserService userService;
```
해당 클래스 내에서는 UserService를 사용하는 곳이 없어 보입니다. UserService를 선언한 이유가 따로 있나요?

yunah7 replied:
코드를 여러군데서 수정하다보니 미처 발견하지 못했습니다 😢 수정하겠습니다 !

uri010 commented on src/main/java/study/sharedcalendar/mapper/ConnectionMapper.java:
```diff
@@ -0,0 +1,19 @@
+public interface ConnectionMapper {
+	String getInviteUrlCode(int id);
+
+	Connection getConnection(@Param("connectorId") int connectorId,
```
Param 어노테이션으로 지정해주지 않아도 myBatis에서 자동으로 connectorId는 #{connectorId}에 매핑해주는 걸로 알고 있습니다! 해당 어노테이션을 제외하고도 한번 실행시켜보세요:)

yunah7 replied:
매개변수가 하나일 경우에는 정상적으로 인식이 되지만, 같은 타입의 매개변수가 2개 이상인 경우에는, 변수명을 맞추어줘도 인식이 되지 않아 Param을 붙여두었습니다! 해당 부분과 관련해 제가 직접 건드린 부분은 properties의 camel-case와 관련된 부분인데, 해당 설정의 true/false와는 관련없이 이러한 에러가 발생해, 설정 관련 빼먹은 부분이 있는지 다시 확인해보겠습니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -0,0 +1,31 @@
+		return connectionService.getInviteUrl();
+	}
+
+	@PostMapping("/change-code")
+	public void changeInviteUrlCode() {
+		connectionService.changeInviteCode();
+	}
```
change에 해당하는 http 메소드가 따로 있을 것 같네요 ! 이 역시 change를 생략할 수 있겠죠

yunah7 replied:
Body로 전달되는 데이터가 없고, 독립적인 데이터의 변경이라 put과 patch 중 고민 중이었는데, 해당 코드에선 Mapping의 타입이 의미적인 부분을 제외하고는 큰 역할을 하지 않는 것 같아, patchMapping으로 변경하겠습니다!

yunah7 committed:
#32 fix : mapping 변경

수정 메소드를 PatchMapping으로 변경
사용하지 않는 객체 삭제

yunah7 commented:
## 변경 내용

**1. Mapping 어노테이션 변경**
```java
	@GetMapping
	public String getInviteUrl() {
		return connectionService.getInviteUrl();
	}

	@PatchMapping
	public void changeInviteUrlCode() {
		connectionService.changeInviteCode();
	}

	@PutMapping("{connectorCode}")
	public void createConnection(@PathVariable String connectorCode) {
		connectionService.createConnection(connectorCode);
	}
```
- 기존 PostMapping으로 통일했던 부분을 기능에 맞는 Mapping으로 수정

**2. 사용하지 않는 객체 삭제**

**3. ConnectionConstant.class**
- 상위 이슈에서 Constant 부분을 #32와 다르게 관심사에 따라 분리해두어, properties 수정 사항에 따라 해당 이슈에서도 이 부분을 추가해두었습니다. 이 외 부분은 merge 시 충돌을 최소화하기 위해 상위 이슈의 내용을 따라가려합니다!

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -0,0 +1,33 @@
+		return connectionService.getInviteUrl();
+	}
+
+	@PatchMapping
+	public void changeInviteUrlCode() {
+		connectionService.changeInviteCode();
+	}
```
`PATCH /connection`만 가지고서 어떤 기능을 하는 API인지 확실하게 알기 어려운 것 같아요.

yunah7 committed:
#33 fix : api명 수정

url을 받아오는 api에 "url" 추가

yunah7 commented:
### 변경 사항
**1. API명 수정**
- 초대 코드를 얻어오는 메소드, 초대 코드를 변경하는 메소드에 "/url" 추가
- Getmapping "~/connections"은 connection을 얻어오는 메소드에 사용되기 때문에, url을 얻어오는 Getmapping은 "/url"을 추가하도록 함

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -0,0 +1,33 @@
+public class ConnectionController {
+	private final ConnectionService connectionService;
+
+	@GetMapping("url")
+	public String getInviteUrl() {
+		return connectionService.getInviteUrl();
+	}
+
+	@PatchMapping("url")
+	public void changeInviteUrlCode() {
+		connectionService.changeInviteCode();
+	}
+
+	@PutMapping("{connectorCode}")
+	public void createConnection(@PathVariable String connectorCode) {
+		connectionService.createConnection(connectorCode);
+	}
```
보통 Mapping 어노테이션의 속성에는 `/url`과 같이 작성합니다.

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -25,7 +26,7 @@ public class UserController {
 	private final LoginService loginService;
 
-	@PostMapping("/sign-up")
+	@PutMapping("/sign-up")
 	public void signUp(@RequestBody @Valid SignUpReq signUpReq) {
 		userService.signUp(signUpReq);
 	}
```
이건 왜 PUT이죠?

sunwootest commented on src/main/java/study/sharedcalendar/service/ConnectionService.java:
```diff
@@ -0,0 +1,50 @@
+		}
+
+		Connection connection = connectionMapper.getConnection(loginId, connectorId);
+		if (connection != null) {
+			if (connection.isActivate()) {
+				throw new DuplicateException(ErrorCode.CONNECT_DUPLICATE);
+			} else {
+				connectionMapper.modifyActivate(connection.getId(), true);
+			}
+		}
+
+		connectionMapper.createConnection(loginId, connectorId);
```
`connection == null`이어도 `connectionMapper.createConnection(loginId, connectorId);`가 실행되겠네요? 좋은 코드 구조는 if문 안에 return혹은 throw와 같은 탈출문을 추가하고 정상적인 동작은 if문 아래에 빼는 구조입니다 :)

sunwootest reviewed (commented):
컨벤션의 변화 때문이었을까요? 변경된 파일 수가 매우 많아졌네요
